packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-listener-action.ts

```diff
@@ -0,0 +1,155 @@
+ * *(weighted)forward*, *fixedResponse* or *redirect* action.
+ *
+ * If an action supports chaining, the next action can be indicated
+ * by passing it in the `next` property.
+ */
+export class NetworkListenerAction implements IListenerAction {
+  /**
+   * Forward to one or more Target Groups
+   */
+  public static forward(targetGroups: INetworkTargetGroup[], options: NetworkForwardOptions = {}): NetworkListenerAction {
+    if (targetGroups.length === 0) {
+      throw new Error('Need at least one targetGroup in a NetworkListenerAction.forward()');
+    }
+    if (targetGroups.length === 1 && options.stickinessDuration === undefined) {
+      // Render a "simple" action for backwards compatibility with old templates
+      return new TargetGroupListenerAction(targetGroups, {
+        type: 'forward',
+        targetGroupArn: targetGroups[0].targetGroupArn,
+      });
+    }
+
+    return new TargetGroupListenerAction(targetGroups, {
+      type: 'forward',
+      forwardConfig: {
+        targetGroups: targetGroups.map(g => ({ targetGroupArn: g.targetGroupArn })),
+        targetGroupStickinessConfig: options.stickinessDuration ? {
+          durationSeconds: options.stickinessDuration.toSeconds(),
+          enabled: true,
+        } : undefined,
+      },
+    });
+  }
+
+  /**
+   * Forward to one or more Target Groups which are weighted differently
+   */
+  public static weightedForward(targetGroups: NetworkWeightedTargetGroup[], options: NetworkForwardOptions = {}): NetworkListenerAction {
+    if (targetGroups.length === 0) {
+      throw new Error('Need at least one targetGroup in a NetworkListenerAction.weightedForward()');
+    }
+
+    return new TargetGroupListenerAction(targetGroups.map(g => g.targetGroup), {
+      type: 'forward',
+      forwardConfig: {
+        targetGroups: targetGroups.map(g => ({ targetGroupArn: g.targetGroup.targetGroupArn, weight: g.weight })),
+        targetGroupStickinessConfig: options.stickinessDuration ? {
+          durationSeconds: options.stickinessDuration.toSeconds(),
+          enabled: true,
+        } : undefined,
+      },
+    });
+  }
+
+  /**
+   * Create an instance of NetworkListenerAction
+   *
+   * The default class should be good enough for most cases and
+   * should be created by using one of the static factory functions,
+   * but allow overriding to make sure we allow flexibility for the future.
+   */
+  protected constructor(private readonly actionJson: CfnListener.ActionProperty, protected readonly next?: NetworkListenerAction) {
+  }
+
+  /**
+   * Render the actions in this chain
+   */
+  public renderActions(): CfnListener.ActionProperty[] {
+    return this.renumber([this.actionJson, ...this.next?.renderActions() ?? []]);
+  }
+
+  /**
+   * Called when the action is being used in a listener
+   */
+  public bind(scope: Construct, listener: INetworkListener) {
+    // Empty on purpose
+    Array.isArray(scope);
+    Array.isArray(listener);
+  }
```

Comment:
Sorry if I'm repeating myself.

> Because it is empty on purpose, for most of the actions

I see only one class implementing it. (is my browser search not finding others?)
Not strongly opposed to what you have here, but making it abstract is the correct (and cheap) option, no?

Reply:
There are actually 2 implementations. The base one, which is simple and does nothing, and a single subclass which implements bind.

The base class is not abstract like you probably assumed it was, it's just a simple data holder (which serves most of the implementations fine, except for the TargetGroup forward action, which needs the bind).

I was saving on a class. Guess I shouldn't do that?

Reply:
Okay I tried making the base class not a simple data carrier, but completely abstract, and it immediately became super-annoying to implement for the Cognito action in the other package, because that package loses access to the "simple data carrying action class".

I did

```ts
class SimpleListenerAction extends ListenerAction {
...
}
```

So I didn't EXPORT it. I *could* export it, but then we have both `ListenerAction` and `SimpleListenerAction` visible to users, which I feel is confusing. So I'm going to keep it as-is.